projects/core/directives/hint/hint.component.ts

```diff
@@ -0,0 +1,91 @@
+import {AnimationOptions} from '@angular/animations';
+import {
+    ChangeDetectionStrategy,
+    Component,
+    ElementRef,
+    HostBinding,
+    Inject,
+    Optional,
+} from '@angular/core';
+import {
+    tuiClamp,
+    TuiContextWithImplicit,
+    TuiDestroyService,
+    TuiHoveredService,
+    tuiPx,
+} from '@taiga-ui/cdk';
+import {TuiDriver, TuiRectAccessor} from '@taiga-ui/core/abstract';
+import {tuiFadeIn} from '@taiga-ui/core/animations';
+import {TuiModeDirective} from '@taiga-ui/core/directives/mode';
+import {TuiHint} from '@taiga-ui/core/interfaces';
+import {TuiPositionService} from '@taiga-ui/core/services';
+import {TUI_ANIMATION_OPTIONS} from '@taiga-ui/core/tokens';
+import {TuiPoint} from '@taiga-ui/core/types';
+import {POLYMORPHEUS_CONTEXT, PolymorpheusContent} from '@tinkoff/ng-polymorpheus';
+import {Observable} from 'rxjs';
+import {takeUntil} from 'rxjs/operators';
+
+import {TuiHintHoverDirective} from './hint-hover.directive';
+import {TuiHintPointerDirective} from './hint-pointer.directive';
+
+@Component({
+    selector: `tui-hint`,
+    template: `
+        <ng-container *polymorpheusOutlet="content as text">{{ text }}</ng-container>
+    `,
+    styleUrls: [`./hint.style.less`],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    providers: [TuiDestroyService, TuiPositionService, TuiHoveredService],
+    animations: [tuiFadeIn],
+})
+export class TuiHintComponent {
+    @HostBinding(`@tuiFadeIn`)
+    readonly animation = {value: ``, ...this.options} as const;
+
+    @HostBinding(`attr.data-appearance`)
+    readonly appearance = this.context.$implicit.appearance || this.mode?.mode;
+
+    @HostBinding(`class._untouchable`)
+    readonly untouchable = this.driver instanceof TuiHintPointerDirective;
+
+    constructor(
+        @Inject(TuiHoveredService) hovered$: Observable<boolean>,
+        @Inject(TuiPositionService) position$: Observable<TuiPoint>,
+        @Inject(TuiDestroyService) destroy$: Observable<void>,
+        @Inject(TuiRectAccessor) protected readonly accessor: TuiRectAccessor,
+        @Inject(ElementRef) private readonly elementRef: ElementRef<HTMLElement>,
+        @Inject(TUI_ANIMATION_OPTIONS) private readonly options: AnimationOptions,
+        @Inject(POLYMORPHEUS_CONTEXT)
+        private readonly context: TuiContextWithImplicit<TuiHint>,
+        @Inject(TuiDriver) private readonly driver: TuiDriver,
+        @Optional()
+        @Inject(TuiModeDirective)
+        private readonly mode: TuiModeDirective | null,
+    ) {
+        position$.pipe(takeUntil(destroy$)).subscribe(([top, left]) => {
+            this.update(top, left);
+        });
+
+        if (driver instanceof TuiHintHoverDirective) {
+            hovered$.pipe(takeUntil(destroy$)).subscribe(hover => driver.toggle(hover));
+        }
+    }
+
+    get content(): PolymorpheusContent {
+        return this.context.$implicit.content;
+    }
+
+    private update(top: number, left: number): void {
```

Comment:
let's add tuiPure to reduce style recalculations

Reply:
It is not pure, it depends on ClientRect values we get in the method.

Reply:
hmm, got it
but for what cases we need to update ClientRect values without updated (top, left) ? I tried to add pure and didn't find that something breaks

Reply:
You are right, we only need these `ClientRect` values to position the beak and for the same `top` and `left` it results in the same position. Nice catch, added `@tuiPure`.